Library/OcVirtualFsLib/VirtualDir.c

```diff
@@ -0,0 +1,586 @@
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+**/
+
+#include <Uefi.h>
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/OcGuardLib.h>
+#include <Library/OcFileLib.h>
+#include <Library/OcVirtualFsLib.h>
+
+#include <Guid/FileInfo.h>
+
+#include "VirtualFsInternal.h"
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirOpen (
+  IN  EFI_FILE_PROTOCOL       *This,
+  OUT EFI_FILE_PROTOCOL       **NewHandle,
+  IN  CHAR16                  *FileName,
+  IN  UINT64                  OpenMode,
+  IN  UINT64                  Attributes
+  )
+{
+  EFI_STATUS         Status;
+  VIRTUAL_DIR_DATA   *Data;
+
+  ASSERT (This != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  if (Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Open (
+      Data->UnderlyingProtocol,
+      NewHandle,
+      FileName,
+      OpenMode,
+      Attributes
+      );
+    if (!EFI_ERROR (Status)) {
+      return CreateRealFile (*NewHandle, NULL, TRUE, NewHandle);
+    }
+    return Status;
+  }
+
+  //
+  // Cannot currently handle any virtualized files in directory.
+  //
+  return EFI_NOT_FOUND;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirClose (
+  IN EFI_FILE_PROTOCOL  *This
+  )
+{
+  EFI_STATUS         Status;
+  VIRTUAL_DIR_DATA   *Data;
+
+  ASSERT (This != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  if (Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Close (
+      Data->UnderlyingProtocol
+    );
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  VirtualDirFree (This);
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirDelete (
+  IN EFI_FILE_PROTOCOL  *This
+  )
+{
+  EFI_STATUS         Status;
+
+  Status = VirtualDirClose (This);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Virtual directories cannot be deleted.
+  //
+  return EFI_WARN_DELETE_FAILURE;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirRead (
+  IN EFI_FILE_PROTOCOL        *This,
+  IN OUT UINTN                *BufferSize,
+     OUT VOID                 *Buffer
+  )
+{
+  EFI_STATUS            Status;
+  VIRTUAL_DIR_DATA      *Data;
+  LIST_ENTRY            *DirEntryLink;
+  VIRTUAL_DIR_ENTRY     *DirEntry;
+  UINTN                 ReadSize;
+  UINTN                 FileStrSize;
+
+  ASSERT (This != NULL);
+  ASSERT (BufferSize != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  //
+  // If our entry position is zero, try to read underlying protocol first.
+  //
+  ReadSize = *BufferSize;
+  if (Data->CurrentEntry == NULL && Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Read (
+      Data->UnderlyingProtocol,
+      BufferSize,
+      Buffer
+      );
+
+      if (EFI_ERROR (Status) || *BufferSize != 0) {
```

Comment:
Please leave a comment on `*BufferSize != 0` check. I guess it is about routing to the original function till we have data from it.